### PR TITLE
s2s backing & issuing using sdk

### PR DESCRIPTION
**Pull request by xiaoch05**

### CODE REVIEW

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        address origin_sender,
+        uint256 amount
+    ) external onlyMessageHandle whenNotPaused {
+        LockedInfo memory lockedMessage = lockedMessages[messageId];
+        require(lockedMessage.hasRefundForFailed == false, "Backing: the locked message has been refund");
+        bytes32 messageHash = hash(abi.encodePacked(messageId, token, origin_sender, amount));
+        require(lockedMessage.hash == messageHash, "Backing: message is not matched");
+        lockedMessages[messageId].hasRefundForFailed = true;
+        // send token
+        require(IERC20(token).transfer(origin_sender, amount), "Backing:unlock transfer failed");
```
Comment (hujw77) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol:203]: If all the `token` deposit to `guard`, address(this)'s balance is 0 ?

**Review by hujw77 — COMMENTED** [review]
LGTM

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/WETH.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.4.18;
+
+contract WETH9 {
+    string public name     = "Wrapped Ether";
+    string public symbol   = "WETH";
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/WETH.sol:7]: Should we rename it to WCRAB / WRING ?

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMappingTokenFactory.sol [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    function hash(bytes memory value) public pure returns (bytes32) {
+        return sha256(value);
+    }
+
+    function rescueFunds(
+        address token,
+        address recipient,
+        uint256 amount
+    ) external onlyAdmin {
+        IERC20(token).transfer(recipient, amount);
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMappingTokenFactory.sol:279]: Some tokens may not be compatible with interface IERC20

File: helix-contract/contracts/mapping-token/v2/Backing.sol [hunk 4]
```diff
@@ -7,6 +7,14 @@ import "./AccessController.sol";
 contract Backing is AccessController, Initializable {
     address public messageHandle;
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/Backing.sol:8]: what it message handle?

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        require(IERC20(token).transferFrom(msg.sender, address(this), amount), "Backing:transfer tokens failed");
+        uint256 balanceAfter = IERC20(token).balanceOf(address(this));
+        require(balanceBefore.add(amount) == balanceAfter, "Backing:Transfer amount is invalid");
+        bytes memory issueMappingToken = abi.encodeWithSelector(
+            IErc20MappingTokenFactory.issueMappingToken.selector,
+            token,
+            recipient,
+            amount
+        );
+        uint256 messageId = _sendMessage(
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol:148]: what is messageId? chain id  + lane id + message nonce?

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        require(lockedMessage.hasRefundForFailed == false, "Backing: the locked message has been refund");
+        bytes32 messageHash = hash(abi.encodePacked(messageId, token, origin_sender, amount));
+        require(lockedMessage.hash == messageHash, "Backing: message is not matched");
+        lockedMessages[messageId].hasRefundForFailed = true;
+        // send token
+        require(IERC20(token).transfer(origin_sender, amount), "Backing:unlock transfer failed");
+        emit TokenUnlockedForFailed(messageId, token, origin_sender, amount);
+    }
+
+    function withdrawFailedTransfer(
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol:208]: Confused naming

Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol:208]:
```suggestion
    function remoteHandleUnlockFailure(
```

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        emit TokenUnlocked(token, recipient, amount);
+    }
+
+    /**
+     * @notice this will be called by messageHandle when the remote issue failed and want to unlock the original token
+     * @param token the original token address
+     * @param origin_sender the origin_sender who will receive the unlocked token
+     * @param amount amount of the unlocked token
+     */
+    function handleWithdrawFailedTransfer(
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol:192]:
```suggestion
    function handleIssuingFailureFromRemote(
```

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/WETH.sol (hunk 2 above)
Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/WETH.sol:7]: This is for testing, the official contract needs to consider this Name and Decimals

File: helix-contract/contracts/mapping-token/v2/Backing.sol (hunk 4 above)
Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v2/Backing.sol:8]: It's the messageEndpoint we have discussed.

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol (hunk 5 above)
Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol]: Now fixed by commit https://github.com/helix-bridge/contracts/pull/7/commits/51817b8b15493a56a4967499b732972e1974c81a
and use laneId + messageNonce

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol (hunk 6 above)
Comment (xiaoch05) [reply helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubBacking.sol]: fixed by https://github.com/helix-bridge/contracts/pull/7/commits/51817b8b15493a56a4967499b732972e1974c81a

**Comment by xiaoch05** [comment]
@HackFisher Please help to review again

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMappingTokenFactory.sol [hunk 8]
```diff
@@ -156,13 +156,13 @@ contract Erc20Sub2SubMappingTokenFactory is DailyLimit, IErc20MappingTokenFactor
         require(mappingToken != address(0), "MappingTokenFactory:mapping token has not created");
         require(amount > 0, "MappingTokenFactory:can not receive amount zero");
         expendDailyLimit(mappingToken, amount);
-        uint256 messageId = IHelixSub2SubMessageHandle(messageHandle).latestRecvMessageId() + 1;
-        require(BitMaps.get(issueMessages, messageId) == false, "message has been accepted");
-        BitMaps.set(issueMessages, messageId);
+        uint256 transferId = IHelixSub2SubMessageHandle(messageHandle).latestRecvMessageId() + 1;
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMappingTokenFactory.sol:159]: What if collision of `transferId` after message handle changed?

File: helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        outboundLanes[outboundLaneId] = nonce;
+        remoteHelix.call(recv);
+        return encodeTransferId(outboundLaneId, nonce);
+    }
+
+    function recvMessage(address receiver, bytes calldata message) external onlyRemoteHelix {
+        require(hasRole(CALLER_ROLE, receiver), "MockS2sMessageHandle:receiver is not caller");
+        // don't make sure this success to simulate the failed case.
+        receiver.call(message);
+        inboundLanes[inboundLaneId] = inboundLanes[inboundLaneId] + 1;
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol:57]: confuse here.

File: helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    function encodeTransferId(bytes4 laneId, uint64 nonce) public pure returns(uint256) {
+        return (uint256(uint32(laneId)) << 64) + uint256(nonce);
+    }
+
+    function decodeTransferId(uint256 transferId) public pure returns(bytes4, uint64) {
+        return (bytes4(uint32(transferId >> 64)), uint64(transferId & 0xffffffffffffffff));
+    }
+
+    function isMessageTransfered(uint256 transferId) public view returns(bool) {
+        (bytes4 laneId, uint64 nonce) = decodeTransferId(transferId);
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol:77]: It looks like that `transferId` is not constructed from `laneId` and `nonce`.

File: helix-contract/contracts/mapping-token/v2/message-handle/DarwiniaSub2SubMessageHandle.sol [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        uint64 latestNonce = SmartChainXLib.lastDeliveredNonce(
+            storageAddress,
+            dstStorageKeyForLastDeliveredNonce,
+            laneId
+        );
+        return nonce <= latestNonce;
+    }
+
+    function latestRecvMessageId() public view returns(uint256) {
+        return SmartChainXLib.lastDeliveredNonce(
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/message-handle/DarwiniaSub2SubMessageHandle.sol:146]: The returned `lastDeliveredNonce ` is `laneId + nonce`, then the naming is a problem ?

File: helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol [hunk 12]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import "../v2/AccessController.sol";
+
+contract MockSub2SubMessageHandle is AccessController {
+    address remoteHelix;
+    mapping(bytes4=>uint64) outboundLanes;
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/test/MockSub2SubMessageHandle.sol:8]:
```suggestion
    mapping(bytes4=>uint64) outboundMessages;
```

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMigration.sol [hunk 13]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.10;
+import "@zeppelin-solidity-4.4.0/contracts/utils/math/SafeMath.sol";
+import "../../interfaces/IERC20.sol";
+
+contract Erc20Sub2SubMigration {
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMigration.sol:7]: should we add an function for burn?

**Review by HackFisher — APPROVED** [review]

File: helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMigration.sol (hunk 13 above)
Comment (HackFisher) [reply helix-contract/contracts/mapping-token/v2/erc20-mapping-protocol/Erc20Sub2SubMigration.sol:7]: Ref:

https://github.com/darwinia-network/darwinia-messages-sol/blob/master/contracts/tokens/contracts/ds-token/token.sol#L427

**Comment by HackFisher** [comment]
Related #9 #7

File: helix-contract/deploy/main.js [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+    const backingNetworkId = "0x0000002d"; //45
+    const backingBridgeNetworkId = "0x70616772";
+    const backingTransactCallIndex = 6657;//1a01
+    const backingSendmsgIndex = 4355;//0x1103
+    const backingOutboundLaneId = "0x726f6c69";
+    const backingStorageKeyForMarketFee = "0x30d35416864cf657db51d3bc8505602f2edb70953213f33a6ef6b8a5e3ffcab2";
+    const backingStorageKeyForLatestNonce = "0xd86d7f611f4d004e041fda08f633f10196c246acb9b55077390e3ca723a0ca1f";
+    const backingStorageKeyForLastDeliveredNonce = "0xd86d7f611f4d004e041fda08f633f101e5f83cf83f2127eb47afdc35d6e43fab";
+
+    const mtfUrl = "https://pangolin-rpc.darwinia.network";
```
Comment (wuminzhe) [inline helix-contract/deploy/main.js:57]: what is `mtf` short for?

Comment (xiaoch05) [reply helix-contract/deploy/main.js:57]: Yes

File: helix-contract/contracts/mapping-token/darwinia/MappingERC20.sol [hunk 15]
```diff
@@ -84,8 +84,8 @@ contract MappingERC20 is IERC20, Ownable, Initializable {
         _mint(account, amount);
     }
 
-    // only factory contract can burn with the authority from users
-    function burn(address account, uint256 amount) external onlyOwner {
+    function burn(address account, uint256 amount) external {
+        require(account == msg.sender, "ERC20: burn address not self");
```
Comment (HackFisher) [inline helix-contract/contracts/mapping-token/darwinia/MappingERC20.sol:88]: Maybe we should add approve feature like dstoken.

Comment (HackFisher) [reply helix-contract/contracts/mapping-token/darwinia/MappingERC20.sol:88]:
```
function burn(address guy, uint wad) public auth stoppable {
        if (guy != msg.sender && _approvals[guy][msg.sender] != uint(-1)) {
            _approvals[guy][msg.sender] = sub(_approvals[guy][msg.sender], wad);
        }

        _balances[guy] = sub(_balances[guy], wad);
        _supply = sub(_supply, wad);
        emit Burn(guy, wad);
    }
```

Comment (HackFisher) [reply helix-contract/contracts/mapping-token/darwinia/MappingERC20.sol:88]: But we did not support auth here, might not easy to use owner and  support auth here.

It is trade off here.

Comment (HackFisher) [inline helix-contract/contracts/mapping-token/darwinia/MappingERC20.sol:88]:
```suggestion
        require(account == msg.sender || owner() == msg.sender, "ERC20: burn address not self or owner");
```

**Pull request merged**